### DISCUSSION TITLE
Update notes

### DISCUSSION THREAD

**Discussion by lshqqytiger** [lshqqytiger/stable-diffusion-webui-directml#191] [Announcements]
# 2023.07.17

- Add ONNX support. (`--onnx`)
Not recommended due to poor performance.
- Load Olive-optimized model when webui started.
- Collect garbage when changing model (ONNX/Olive).
- Some minor changes.

## Expected errors & solutions

1.
```
RuntimeError: Couldn't install onnxruntime.
Command: python -m pip install --no-deps onnxruntime --prefer-binary
Error code: 1
stdout: Collecting onnxruntime
  Using cached onnxruntime-1.15.1-cp310-cp310-win_amd64.whl (6.7 MB)
Installing collected packages: onnxruntime

stderr: ERROR: Could not install packages due to an OSError: [WinError 5]
Consider using the `--user` option or check the permissions.
```

Solution:
```
activate venv.
(venv) $ pip uninstall onnxruntime
(venv) $ pip uninstall onnxruntime-directml
restart webui.
```

2.
```
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.5
    Uninstalling numpy-1.23.5:
      Successfully uninstalled numpy-1.23.5

stderr: ERROR: Could not install packages due to an OSError: [WinError 5]
Consider using the `--user` option or check the permissions.
```

Solution:
```
activate venv.
(venv) $ pip install numpy==1.23.4
start webui with --skip-install.
```

**Comment by lshqqytiger** [comment]
# 2023.07.19

- Added ONNX Runtime tab in Settings.
- Added `Reload model before each generation.` option.
`Reload model before each generation.` unloads and loads model before each generation.
With `Reload model before each generation.`:
![reload](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/39524005/08010768-5ce3-4e87-899c-7a802c5ab4fa)
Without `Reload model before each generation.`:
![load_once](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/39524005/a6e98fa8-c5fa-43d2-af1f-06110d779f35)
- Automatically calculate dimension.
- You can change sampler when using optimized model.
- You can select which sampler will be used (ONNX & Olive).
- Implemented SDXL optimization. (couldn't test. it requires more than 32GB of ram)
- Add `offload state dict` in ONNX settings.

# Critical changes

- `diffusers` should be updated.
```
(venv) $ pip install --upgrade diffusers
```
- `invisible-watermark` should be installed.
```
(venv) $ pip install invisible-watermark
```

**Reply by Rizzlord** [reply]
what does "offload state dict" do? also, when i generate 4 images the vram wont get cleared, although i checked the option. Only when i Generate one by one image.

**Reply by lshqqytiger** [reply]
1. `Offload state dict` temporarily offloads the CPU state dict to the hard drive to avoid running out of CPU RAM if the weight of the CPU state dict + the biggest shard of the checkpoint does not fit. (from the comments of diffusers.DiffusionPipeline)
2. Should be fixed in 2023.07.24.

**Reply by Rizzlord** [reply]
2023-07-24 14:06:01.0906015 [W:onnxruntime:, session_state.cc:1171 onnxruntime::VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments. Got this multiple times, any suggestions?

**Reply by lshqqytiger** [reply]
Most of warning messages are ignorable. If optimized model generates incoherent images or model fails to optimize, report me with logs.

**Reply by sgumil** [reply]
Just tried to load SDXL base safetensors but size mismatch

**Reply by lshqqytiger** [reply]
SDXL is currently not supported by Olive. I just implemented myself. If you attach log and traceback, I will have a look.

**Comment by lshqqytiger** [comment]
# 2023.07.24

- Fix `Reload model before each generation` behavior.
- Fix Olive model loading bug.

**Comment by lshqqytiger** [comment]
# 2023.08.01

- Add new option: DirectML memory stats provider.
1. Performance Counter. (default)
    Get vram size allocated to & used by python.exe from pdh.dll.
    Generation can be slower than atiadlxx.
    Use memory less greedy than atiadlxx.
    Windows only.
2. atiadlxx.
    Get max vram size and available vram size from AMD GPU driver (atiadlxx.dll).
    Use memory more greedy than Performance Counter.
    for AMD GPU only.
    Windows & WSL are supported.
3. None.
    Assume available vram size is 8GB.
    Use memory regardless of current vram usage.

**Reply by Rizzlord** [reply]
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ E:\StableDiffusion\launch.py:39 in <module>                                                      │
│                                                                                                  │
│   36                                                                                             │
│   37                                                                                             │
│   38 if __name__ == "__main__":                                                                  │
│ ❱ 39 │   main()                                                                                  │
│   40                                                                                             │
│                                                                                                  │
│ E:\StableDiffusion\launch.py:35 in main                                                          │
│                                                                                                  │
│   32 │   if args.test_server:                                                                    │
│   33 │   │   configure_for_tests()                                                               │
│   34 │                                                                                           │
│ ❱ 35 │   start()                                                                                 │
│   36                                                                                             │
│   37                                                                                             │
│   38 if __name__ == "__main__":                                                                  │
│                                                                                                  │
│ E:\StableDiffusion\modules\launch_utils.py:443 in start                                          │
│                                                                                                  │
│   440                                                                                            │
│   441 def start():                                                                               │
│   442 │   print(f"Launching {'API server' if '--nowebui' in sys.argv else 'Web UI'} with argum   │
│ ❱ 443 │   import webui                                                                           │
│   444 │   if '--nowebui' in sys.argv:                                                            │
│   445 │   │   webui.api_only()                                                                   │
│   446 │   else:                                                                                  │
│                                                                                                  │
│ E:\StableDiffusion\webui.py:54 in <module>                                                       │
│                                                                                                  │
│    51 startup_timer.record("import ldm")                                                         │
│    52                                                                                            │
│    53 from modules import extra_networks                                                         │
│ ❱  54 from modules.call_queue import wrap_gradio_gpu_call, wrap_queued_call, queue_lock  # noq   │
│    55                                                                                            │
│    56 # Truncate version number of nightly/local build of PyTorch to not cause exceptions with   │
│    57 if ".dev" in torch.__version__ or "+git" in torch.__version__:                             │
│                                                                                                  │
│ E:\StableDiffusion\modules\call_queue.py:6 in <module>                                           │
│                                                                                                  │
│     3 import threading                                                                           │
│     4 import time                                                                                │
│     5                                                                                            │
│ ❱   6 from modules import shared, progress, errors                                               │
│     7                                                                                            │
│     8 queue_lock = threading.Lock()                                                              │
│     9                                                                                            │
│                                                                                                  │
│ E:\StableDiffusion\modules\shared.py:95 in <module>                                              │
│                                                                                                  │
│    92                                                                                            │
│    93                                                                                            │
│    94 if cmd_opts.backend == 'directml':                                                         │
│ ❱  95 │   directml_do_hijack()                                                                   │
│    96                                                                                            │
│    97                                                                                            │
│    98 def reload_hypernetworks():                                                                │
│                                                                                                  │
│ E:\StableDiffusion\modules\dml\__init__.py:69 in directml_do_hijack                              │
│                                                                                                  │
│   66 │   │   │   lambda orig_func, *args, **kwargs: orig_func(args[0].astype('float32')),        │
│   67 │   │   │   lambda *args, **kwargs: args[1].dtype == float)                                 │
│   68 │                                                                                           │
│ ❱ 69 │   _set_memory_provider()                                                                  │
│   70                                                                                             │
│   71 def directml_override_opts():                                                               │
│   72 │   _set_memory_provider()                                                                  │
│                                                                                                  │
│ E:\StableDiffusion\modules\dml\__init__.py:14 in _set_memory_provider                            │
│                                                                                                  │
│   11 do_nothing = lambda: None                                                                   │
│   12                                                                                             │
│   13 def _set_memory_provider():                                                                 │
│ ❱ 14 │   from modules.shared import opts, cmd_opts, log                                          │
│   15 │                                                                                           │
│   16 │   if opts.directml_memory_provider == "Performance Counter":                              │
│   17 │   │   from .backend import pdh_mem_get_info                                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ImportError: cannot import name 'opts' from partially initialized module 'modules.shared' (most likely due to a circular
import) (E:\StableDiffusion\modules\shared.py) 

Tried to make a fresh venv, same error.

**Reply by AvengedPotat** [reply]
Same error here as well

**Reply by Miraihi** [reply]
Same error unfortunately. [Pastebin traceback](https://pastebin.com/Dm0EHj3U).
Vladmandic version of your update seems to be loading fine without initial errors.

**Comment by pw405** [comment]
Darn, same here.  Log is relatively long, so I'll just attach as text if it is of any value. 

[Stable Diffusion DirectML 8-1-23 Update Error Log.txt](https://github.com/lshqqytiger/stable-diffusion-webui-directml/files/12223092/Stable.Diffusion.DirectML.8-1-23.Update.Error.Log.txt)

I'm relatively new to Git Hub... can I just do a pull of the prior version?

**Reply by Miraihi** [reply]
I personally use [Github Desktop](https://desktop.github.com/) so I can check the updates anytime and see the full commit history with the option to revert any of them temporarily or switch to the other branch.

Also there's the alternative webui called [SD.Next](https://github.com/vladmandic/automatic) that also has DirectML functionality. At this moment it works with the latest updates from Ishqqytiger.

**Reply by pw405** [reply]
Oh interesting, I may have to try SD.Next!!  But.... not sure if I'm quite ready to go with an entirely new UI just yet.  I'm sure @lshqqytiger will have an update/fix soon.

**Reply by pw405** [reply]
I ended starting over with previous commit, but kept the broken commit untouched/in place.  Instead I cloned prior commit to my secondary drive (Gen 4 NVME) instead of putting it on my C:  drive (Gen 3 NVME.)

Model loading from disk is faster!  I figure if @lshqqytiger  has a new commit soon, I'll just move back to original install.  Would eventually like to have everything on the Gen 4 NVME anyways, so maybe this broken commit is a blessing in disguise?

**Reply by Miraihi** [reply]
The error has been fixed today. Pull the latest commit.

**Reply by pw405** [reply]
Bingo! Jurassic Park is back online!
On behalf of baby kangaroo in a suit at a bar, thank you @lshqqytiger ! 

![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/b722a9bb-38b4-4ee6-b496-30e8c93d8617)

**Comment by lshqqytiger** [comment]
# 2023.08.27

- Implement ONNX/Olive Img2Img. (has many problems..)
- Implement parallelism.
![스크린샷 2023-08-27 173345](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/39524005/8757f618-4a65-4924-9f98-f56f0079f0c2)
▲ my old second gpu works for the text encoder while the first gpu works for the unet and the vae.

**Comment by Rizzlord** [comment]
when i try to optimize sdxl via Olive it gives me this error "TypeError: StableDiffusionPipeline.__init__() got an unexpected keyword argument 'text_encoder_2'" 
Update : It works somehow with Dreamshaper 8 (its a sdxl model) but still no chance with native sdxl

**Reply by lshqqytiger** [reply]
Did you check `Text Encoder 2`? It is required to optimize SDXL model. Otherwise, it will try to load as SD model.

**Comment by lshqqytiger** [comment]
# 2023.10.13

### Every bug reports are welcome!

- New options for ONNX Runtime: `Use Just-in-Time optimization.`, `Cache optimized model.`
- `Use Just-in-Time optimization`: Every unoptimized model will be shown. If the selected model is optimized, it will generate image using the optimized model. If not, the model will be optimized first. After then, until webui lives, optimized model will be maintained.
- `Cache optimized model`: If you uncheck this option, it will remove every cache when you optimize model.
- Improved system memory efficiency.
- Improved graphics memory efficiency. (onnxruntime 0.16)

**Comment by lshqqytiger** [comment]
# 2024.02.07

## Every bug reports are welcome!

- Refactored onnx/olive. for details: https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/149#discussioncomment-8392257

**Reply by pw405** [reply]
Has anybody been able to get this update working yet?  I can't seem to find a solution.

**Comment by lshqqytiger** [comment]
# 2024.02.07

## Every bug reports are welcome!

- Added `--use-zluda`.

## ZLUDA Prerequisites

- The latest ZLUDA release (from [here](https://github.com/lshqqytiger/ZLUDA/releases))
- AMD HIP SDK

**Reply by tocram1** [reply]
So, is there a tutorial somewhere about how to install those pre-requisites? The AMD HIP SDK is pretty much an installation wizard as much as I remember, but is this fork of ZLUDA required somewhere in the `stable-diffusion-webui-directml` or does it need to be installed into our AMD drivers? Thank you in advance for your reply!

**Reply by Drakensangs** [reply]
Would appreciate some further clarification on this subject as well.

**Reply by lshqqytiger** [reply]
https://github.com/vladmandic/automatic/wiki/ZLUDA

**Comment by pw405** [comment]
After yesterday's update, I seem to be having trouble with Image to Image when trying to use Ultimate SD Upscale extension  (https://github.com/Coyote-A/ultimate-upscale-for-automatic1111).  It is used to slice an image in to tiles, so you can upscale far beyond the normal dimensions possible. 

Odd thing, It doesn't crash or generate any error messages.  Once it finishes generating the first tile, it just stops and doesn't move on to the second tile. Interrupting or skipping doesn't work, you have to actually close the console out and re launch. 

Using following command line arg's with 7900 XTX:

set COMMANDLINE_ARGS= --use-directml --medvram --no-half

**Reply by Drakensangs** [reply]
I'm having the same issue with using Hires.fix. When it finishes generating the image and moves on to using Hires.fix, it just completely stops and nothing happens.

**Reply by pw405** [reply]
Interesting, seems like the same issue then.  Question, if I may: even with 24GB of VRAM I've never been able to get HiRes fix to work without throwing out of VRAM error.  Can you successfully use HiRes fix with SDXL models? If so, what settings generally work?

**Reply by lshqqytiger** [reply]
Maybe depends on the sampler. What sampler did you use, and do other samplers hang terminal?
About hires.fix, you'd better to try ZLUDA that utilizes VRAM more efficiently.

**Reply by pw405** [reply]
The issue with Ultimate SD Upscale seemed to happen with a variety of samplers across different SDXL models.  

Would have been DPM++ 2M Karras, DPM++ SDE Karras, or Euler samplers.

I know Ultimate SD Upscale was working as of April 12th.

I want to try Zluda!! Haven't had time to run through an install yet though.

**Reply by Drakensangs** [reply]
Tried hires.fix with multiple samplers and upscalers. The upscaler seems to be the issue for hires.fix, at least for me. Latent works, while every other non-Latent upscaler ends up hanging the terminal.

**Reply by lshqqytiger** [reply]
I'll take a look when I'm free.
For now, you can try ZLUDA or revert to 1.8.

DirectML: slow, use VRAM so much and not efficiently, doesn't return freed VRAM, some decent pytorch features are not supported, many bugs, closed source, don't know about Fast Fourier Transform (FFT) & DNN & Runtime Compiler (RTC).
ZLUDA: fast, use VRAM more efficiently, can return freed VRAM, support decent pytorch features, unstable, can train something, open source, FFT & DNN & RTC disabled.

**Reply by Drakensangs** [reply]
Forgive my ignorance, but how does one go about reverting to 1.8?

**Reply by lshqqytiger** [reply]
run `git reset --hard 8e2f087ab7925fbbddf0cc30d68655bf070e98b2`

**Reply by lshqqytiger** [reply]
I created tag `v1.8.0-amd`. You can `git checkout tags/v1.8.0-amd`.

**Comment by pw405** [comment]
> I'll take a look when I'm free. For now, you can try ZLUDA or revert to 1.8.
> 
> DirectML: slow, use VRAM so much and not efficiently, doesn't return freed VRAM, some decent pytorch features are not supported, many bugs, closed source, don't know about Fast Fourier Transform (FFT) & DNN & Runtime Compiler (RTC). ZLUDA: fast, use VRAM more efficiently, can return freed VRAM, support decent pytorch features, unstable, can train something, open source, FFT & DNN & RTC disabled.

Trying to get ZLUDA up and running.  I installed the AMD HIP SDK and placed the extracted ZLUDA download in my root SD folder:

C:\Users\alias\STABLEDIFFUSION\ZLUDA-windows-amd64

I followed the guide referenced in https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/191#discussioncomment-9066846

I renamed my VENV folder, purged cache (`pip cache purge`), and launched webui-user.bat with:

`set COMMANDLINE_ARGS=  --use-zluda --medvram`

This gives error:  `RuntimeError: Torch is not able to use GPU; add --skip-torch-cuda-test to COMMANDLINE_ARGS variable to disable this check`

Adding ` --skip-torch-cuda-test` results in:

```
Traceback (most recent call last):
  File "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\launch.py", line 48, in <module> main()
  File "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\launch.py", line 39, in main prepare_environment()
  File "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\modules\launch_utils.py", line 658, in prepare_environment from modules.onnx_impl import initialize_olive
  File "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\modules\onnx_impl\__init__.py", line 4, in <module> import torch
  File "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\venv\lib\site-packages\torch\__init__.py", line 141, in <module> raise err
OSError: [WinError 126] The specified module could not be found. Error loading "C:\Users\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\venv\lib\site-packages\torch\lib\caffe2_nvrtc.dll" or one of its dependencies.
```
Any idea what to do?  I wasn't entirely sure if I "added the folders to path" correctly as mentioned here.  
https://github.com/vladmandic/automatic/wiki/ZLUDA#add-folders-to-path

I edited the Windows Environment Variables system variable Path to include %HIP_PATH%bin as shown here:
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/683132f7-f150-4eea-a269-5bb9822a009c)

**Reply by lshqqytiger** [reply]
1. Add path to ZLUDA in PATH
2. remove torch, torchvision using these commands
```
.\venv\Scripts\activate
pip uninstall torch torchvision -y
```
and try again with `--use-zluda` without `--skip-torch-cuda-test`.
You don't need to keep `--medvram` for the most cases.

**Reply by pw405** [reply]
Thanks, will try shortly when back at PC.  Not understanding when you say "Add path to Zluda in PATH"... 

As in, the bin folder of the ROCM install?
C:\Program Files\AMD\ROCm\5.7\bin

Do I add it in the same exact place as my screenshot, so as to REPLACE what is currently showing as %HIP_PATH%bin in previous screenshot?

**Reply by lshqqytiger** [reply]
`%HIP_PATH%bin` has no problem if `%HIP_PATH%` exists and is set to proper ROCm path. (and it should if the installation finished without any problem)
However, you should not have `torch` installed before the first launch.

launch.py behavior:
1. `torch-directml` installed? => skip ZLUDA setup
2. `torch` installed? => skip ZLUDA setup
3. `ZLUDA` environment variable is set? => use `%ZLUDA%` as zluda path
4. one of the paths in PATH has `zluda_redirect.dll` as a child? => use that path as zluda path
5. If zluda path is not found, download zluda from github. ← you may be here.

Therefore, you may have `.zluda` folder in your webui directory.
You can ignore my first comment. It won't help `caffe2_nvrtc.dll` error.

Actually, I haven't met WinError 126 `caffe2_nvrtc.dll`, but I have been reported same errors from the community.
For now, I think it is mostly due to missing or conflicting PATH or dll files.

1. You should have `path\to\rocm\5.7\bin` in PATH. (`%HIP_PATH%bin` if `HIP_PATH` is set correctly)
2. launch.py automatically append `path\to\zluda` to PATH, but you can add it yourself to make sure.
3. You should not have multiple nvcuda.dll, nvml.dll, etc. They make conflicts.
4. You should have Visual C++ Runtime installed.

If you checked all the things and it still doesn't work, follow this:
1. Run
```
.\venv\Scripts\activate
pip uninstall torch -y
pip install torch==2.2.2 torchvision --index-url https://download.pytorch.org/whl/cu118
```
2. Make duplicates of `cublas.dll`, `cusparse.dll`. (you can find them from zluda folder)
3. Rename them to `cublas64_11.dll`, `cusparse64_11.dll`.
5. Copy them to `path\to\webui\venv\Lib\site-packages\torch\lib`. (overwrite existing dlls)

**Reply by pw405** [reply]
Awesome, thanks for detail!  Will try on my lunch break (hopefully). 

And FYI - the ROCm Installer **does** successfully add two entries to System Variables:  HIP_PATH and HIP_PATH_57.
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/b9b65872-9f21-46f6-823e-9450496ba80a)

So, I'm assuming the % signs in ` %HIP_PATH%bin `  are wildcards in the _Path_ System Variable - simply to point it to the bin folder:   
C:\Program Files\AMD\ROCm\5.7\bin

**Reply by pw405** [reply]
We have blastoff! 

I can't say for sure - I think the issue was simply the reinstall of Torch & Torch Vision.  Before I launched with `--use-zluda` the first time, I had no venv folder as far as the script knew.  (I renamed to venvDML before launching).  

Today I ran the command:
```
\venv\Scripts\activate
pip uninstall torch torchvision -y
```
I launched webui with: `set COMMANDLINE_ARGS= --use-zluda`  It successfully downloaded & installed Torch 2.2.2 and torchvision.

I selected  a different model, but it took ~5 minutes to load!!?  WTF? (PCI Gen 3 SSD, Ryzen 5800X).

Assuming it wasn't working, I closed out.  Then I renamed the two DLL files from ZLUDA folder and overwrite the files in  `path\to\webui\venv\Lib\site-packages\torch\lib`

Launched and tried A simple prompt with SDXL model.  Nothing. No errors. But no progress for ~3 minutes.  Then it finished.  **Insanely fast**. I thought I was on a Lightning model. 

For others wanting to try, a summary of things I found to do (7900XTX):

Download & Install AMD HIP SDK.
Download & extract the Zluda folder  (I just put in top Stable Diffusion folder)
![image](https://github.com/lshqqytiger/stable-diffusion-webui-directml/assets/20309609/d4549245-d509-4443-9d2b-013744a51001)

Open Windows Environment Variables, under System Variables, find Path.  Select it, edit, New: create entry with: `%HIP_PATH%bin` as seen in https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/191#discussioncomment-9129229 

Venv folder:  You can delete, or simply rename.  I wanted to keep mine in case I switch back to DirectML.   Consider the uninstall torch/torchvision if you get errors. 

Copy/Paste the DLL's as directed above. 

Upon launching, make sure it mentions it is using ZLUDA in the console:

C:\Users\alias\stable-diffusion-webui-directml>set COMMANDLINE_ARGS= --use-zluda
C:\Users\alias\stable-diffusion-webui-directml>call webui.bat
venv "C:\Users\alias\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
WARNING: ZLUDA works best with SD.Next. Please consider migrating to SD.Next.
**Using ZLUDA in C:\Users\alias\stable-diffusion-webui-directml\.zluda**

Then, be prepared to **WAIT for that first model** load/generation.  There is no sign anything is happening. 

HOLY CRAP its fast though!!  From memory, I know a typical SDXL model  at 1216 X 832  with DirectML is ~2.5 seconds/it (simple samplers: DPM ++ 2M Karras) or ~4 seconds/it with complex samplers (DPM ++ SDE Karras, Restart, etc.). 

ZLUDA Test (JuggernautV9 model):  DPM++ 2M Karras:  ~3 iterations/second. **MASSIVE** Performance boost.  30 step 832 X 1216 image in 12.8 seconds?  WOW!  GPU Hotspot temp never even hit 80°. Common for it to hit ~92° with DirectML. 

Can't wait to test more!

**Reply by DJHanceNL** [reply]
Wow, it's working now. But the wait on first generation is real. I think, 10 percent cpu usage, and it took 15 min. Didnt know what it was doing, but after, everything is back to normal (but 2x speed) 


Edit*

This is so cool, sdxl would give so many errors of out of vram. Now on 832x1216 it's usiing 9gb of 16 (6900xt) 2 inits a sec with dpm ++ 3m SDE Karras,  (It's directml x4 in performance)

**Comment by pw405** [comment]
I accidently ran a GIT PULL while starting today and can't seem to get the UI to launch using Zluda any longer.  I guess I had it off since my last post.

I repeated all steps outlined in my previous post:  https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/191#discussioncomment-9135947  

However I still get the following error:

```
C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml>set COMMANDLINE_ARGS= --use-zluda
C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml>set SAFETENSORS_FAST_GPU=1
C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml>call webui.bat
venv "C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
WARNING: ZLUDA works best with SD.Next. Please consider migrating to SD.Next.
Using ZLUDA in C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\.zluda
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: v1.9.3-amd-9-g9514d919
Commit hash: 9514d9194d6a8a45d3ceb42567e45d020d5226c0
Traceback (most recent call last):
  File "C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\launch.py", line 48, in <module>
    main()
  File "C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\launch.py", line 39, in main
    prepare_environment()
  File "C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\modules\launch_utils.py", line 592, in prepare_environment
    raise RuntimeError(
RuntimeError: Torch is not able to use GPU; add --skip-torch-cuda-test to COMMANDLINE_ARGS variable to disable this check
Press any key to continue . . .
```

**Reply by lshqqytiger** [reply]
Fixed in e2cbdab3375deea621a48de269fd4ac47ff6cece

**Reply by pw405** [reply]
Nice! That did it!

**Comment by pw405** [comment]
After the update today, it says it "failed to install Zluda Path, missing 1 required positional argument: 'zluda_path' "

```
C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml>call webui.bat
venv "C:\alias\STABLEDIFFUSION\WEBUI\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
WARNING: ZLUDA works best with SD.Next. Please consider migrating to SD.Next.
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: v1.9.3-amd-21-g69011ef6
Commit hash: 69011ef6095c5e45d3a548c218fd1b024ea62810
Failed to install ZLUDA: install() missing 1 required positional argument: 'zluda_path'
Using CPU-only torch
Installing requirements
Collecting protobuf<=4.9999,>=4.25.3
  Downloading protobuf-4.25.3-cp310-abi3-win_amd64.whl (413 kB)
```

**Reply by ride5k** [reply]
can confirm, also experienced error.  wiped venv and rebuilt yet issue persists.

**Reply by TiberiumCat** [reply]
Same. I got it working.
I **THINK** what worked was updating zluda to v3.7 (5/17/24)

**Reply by ride5k** [reply]
> Same. I got it working. I **THINK** what worked was updating zluda to v3.8 (5/17/24)

@TiberiumCat i looked around for a precompiled windows version of the latest release but could not find one.  did you compile it yourself?

**Reply by TiberiumCat** [reply]
> @TiberiumCat i looked around for a precompiled windows version of the latest release but could not find one. did you compile it yourself?

lshqqytiger did that for you (link). I litterly just copied the new files into the old folder
https://github.com/lshqqytiger/ZLUDA/releases/tag/rel.4f12e8cfe9fa53bd29637c0bfe6fc88494d2b10f

**Reply by pw405** [reply]
> Same. I got it working. I **THINK** what worked was updating zluda to v3.7 (5/17/24)

Dang, no luck.  I downloaded the 3.7.2 version of Zluda, extracted, renamed two files and pasted them in the 
`...WEBUI\stable-diffusion-webui-directml\venv\Lib\site-packages\torch\lib`  folder.

Launching still gives same error. 

I tried copy/pasting all DLL's from the Zluda download in to the torch\lib folder.  Also same error. 

Failed to install ZLUDA: install() missing 1 required positional argument: 'zluda_path'

I'll keep everybody posted if I figure out it out.  Until then, you can always roll back:  

`git checkout 517aaaff2bb1a512057d88b0284193b9f23c0b47`

**Reply by lshqqytiger** [reply]
Did you git pull?

**Reply by TiberiumCat** [reply]
> > Same. I got it working. I **THINK** what worked was updating zluda to v3.7 (5/17/24)
> 
> Dang, no luck. I downloaded the 3.7.2 version of Zluda, extracted, renamed two files and pasted them in the `...WEBUI\stable-diffusion-webui-directml\venv\Lib\site-packages\torch\lib` folder.

sorry, it was both an hour past bed time and just as I woke up when I last looked at this thread.

I don't mean JUST the torch/lib files. I **also** mean the folder you have the REST of the zluda files. 
If you did it like the guild I followed, it's the folder you set a "path" to in "system environment variables". 
I just did what this guy said the first time. He talks you through "system environment variables" settings at 2:46, you can backtrack to wherever you put the zluda folder from there (shrug emoji)  https://youtu.be/n8RhNoAenvM?si=ZIh5_ysEtDm-j9K4&t=166

The first time I ran SD after doing it, I **think** it still showed the error but I ran a test anyway and it took like 20 minutes to do _whatever_ zluda does the first time. Now it's fine. (shrug emoji) 

I hope this clarifies what I did.

**Reply by pw405** [reply]
The version posted yesterday with the fix is now working successfully @lshqqytiger !

I don't know what I did yesterday to break it so badly!

Hey @lshqqytiger , what does double clicking the Zluda.exe in your download package do? 

I think that might have replaced some files that were incompatible.  I tried to git checkout with the hash in my previous comment, but I was getting errors that Zluda_redirect.dll wasn't working.  

Then I was back to the error about a DLL failing. 

Eventually after wiping venv and rolling back versions in the .Zluda folder, etc, I got the previous version to run.  

This morning, simply doing a git pull on master/newest branch was all that was needed to fix.  

Thanks for the work @lshqqytiger !  You have a Buy Me a Coffee, Patreon, or anything?  I feel like we've all greatly benefited from your work for free. What about an RX 6800?  Ill make you an amazing deal!
